dashboard/modules/job/job_manager.py

```diff
@@ -379,8 +384,11 @@ async def run(
                 if sys.platform == "win32" and self._win32_job_object:
                     win32job.TerminateJobObject(self._win32_job_object, -1)
                 elif sys.platform != "win32":
-                    # TODO (jiaodong): Improve this with SIGTERM then SIGKILL
-                    child_process.kill()
+                    os.killpg(os.getpgid(child_process.pid), signal.SIGTERM)
+                    try:
+                        child_process.wait(self.WAIT_FOR_JOB_TERMINATION_S)
+                    except subprocess.TimeoutExpired:
+                        child_process.kill()
```

Comment:
Nit: does it make sense to pull this out into a helper function to avoid duplicating the code?

Reply:
I've removed the block of code in `_polling`.